File: sources/academy/webscraping/scraping_basics_python/exercises/cnn_sports_shortest_article.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin


def download(url: str) -> BeautifulSoup:
    response = httpx.get(url)
    response.raise_for_status()
    return BeautifulSoup(response.text, "html.parser")


listing_url = "https://edition.cnn.com/sport"
listing_soup = download(listing_url)

results: list[tuple[int, str]] = []
for card in listing_soup.select('.layout__main .card'):
    link = card.select_one('.container__link')
    if not link or 'href' not in link.attrs:
        continue

    article_url = urljoin(listing_url, link['href'])
    article_soup = download(article_url)
    content = article_soup.select_one('.article__content')

    if not content:
        continue

    results.append((len(content.get_text()), article_url))

results.sort()
if results:
    print(results[0][1])

File: sources/academy/webscraping/scraping_basics_javascript/exercises/cnn_sports_shortest_article.mjs
import * as cheerio from 'cheerio';

async function download(url) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  const html = await response.text();
  return cheerio.load(html);
}

const listingUrl = 'https://edition.cnn.com/sport';
const $ = await download(listingUrl);

const results = await Promise.all(
  $('.layout__main .card').toArray().map(async (element) => {
    const $element = $(element);
    const $link = $element.find('a').first();
    if (!$link.length) {
      return null;
    }

    const articleUrl = new URL($link.attr('href'), listingUrl).href;
    const $article = await download(articleUrl);
    const content = $article('.article__content').text().trim();

    if (!content) {
      return null;
    }

    return { url: articleUrl, length: content.length };
  }),
);

const nonEmpty = results.filter((item) => item && item.length > 0);
nonEmpty.sort((a, b) => a.length - b.length);

if (nonEmpty.length > 0) {
  console.log(nonEmpty[0].url);
}
